src/Interpreters/join_common.cpp

```diff
@@ -164,6 +188,45 @@ Columns materializeColumns(const Block & block, const Names & names)
     return materialized;
 }
 
+bool canCastJoinColumns(const Block & left_block, const TableJoin & table_join)
```

Comment:
Looks like what this function does doesn't match its name. We can cast the columns always, if the cast is possible between the corresponding types. But it doesn't check the types?

The number of keys on both sides is probably checked elsewhere, it's not related to the cast.

I'm not sure what the check for `left_block` does at all.

Reply:
Renamed and comment added
https://github.com/ClickHouse/ClickHouse/blob/b4acfc872f47fc209377ac55123e6a7fa73e54f1/src/Interpreters/join_common.cpp#L180-L184